### ISSUE TITLE
Arguments passed to `fin exec` are inconsistent with `fin run-cli`

### ISSUE THREAD

**Issue by GMTA** [docksal/docksal#738]
**Description**
Running Docksal 1.60.2 and docksal/cli:2.3-php7.2.

Using `fin run-cli [args]` works as you expect: a standalone cli container is spawned and `[args]` is fed to it as if you ran `fin bash [new-container]` and entered the command manually.

However, `fin exec [args]` is problematic in that it has its own logic:

If argument `$1` is an existing file, it is passed to `bash -c` _without_ any of the other arguments passed to it. Example:

`fin exec path/to/script.sh arg1 arg2`

Those `arg1` and `arg2` arguments are never passed to `script.sh`. Additionally, if you try to pass a single argument to be executed as a shell statement, it fails because it escapes the entire argument:

`fin exec "source ~/.profile && npm run" something`

The fact that `something` is outside of the quotes is caused by Bash pulling it to the side when using `$@` in a custom command, for example:

`fin exec "source ~/.profile && npm ${@}"`

**Suggested solution**
In my opinion, `fin exec [args]` should behave exactly like `fin bash` followed by entering `[args]` by hand and pressing [enter]. I think this is what people would expect from such a command.

In order to support this, the only thing `fin exec [args]` should do is run a `docker exec` in the right container with `bash -c "[args]"` where `"[args]"` is one big string containing all the remaining arguments passed to `fin exec`. No _additional_ escaping should take place and there should not be a check if a file exists and then run that in a different flow.

My guess is that the `fin exec [file]` format was introduced to help people running scripts that do not have the executable bit set, but it might help to just err out with a message asking the user to set that bit.

**Comment by achekulaev** [comment]
I will actually reopen it for myself to see if I can revisit the way I compose the command in the exec. I might be able to compose it better.

**Comment by achekulaev** [comment]
I was doing this code so long time ago, that I myself forgot the reasons why it is done the way it is done. So I will describe it here for you and for myself.

---

The main reason is: `docker exec` only accepts string. This is ok for programming, but this is not user friendly at all. To print "Hello, World!" using docker exec:

```bash
docker exec bash -ilc 'echo "Hello, World!"'
```

Mind the quotes alternation! I wanted users to be able to do `fin exec ls -la`, without a need to always encapsulate command in quotes `fin exec "ls -la"` like you need to do with `docker exec`. And to print hello world easily `fin exec echo "Hello, World!"` not like `fin exec 'echo "Hello, World!"'`. And that's only the tip of the iceberg, If you need to use variable inside the exec string, it would become even more problematic since you need to break single quotes in this case or alternate them in a different way.

```bash
myvar="World"

# with current fin exec
fin exec echo "Hello, $myvar."

# with fin exec matching docker exec
fin exec "echo 'Hello, ${myvar}'"
```

It would be a hell to a regular user, who would have to remember about all the peculiar things about bash. This is the reason why fin does not just match `docker exec`. 

Instead `fin exec` accepts an array of parameters like any normal bash function or utility. So `$@` is an array. Feeding `$@` directly to `docker exec` will fail to work properly.

```bash
_test() 
{
    docker exec -u docker -it YOURPROJECT_cli_1 bash -ilc "$@"
}

_test ls -la
```

Example above will do `ls` instead of `ls -la`.

So conversion from array `$@` to string passed to `docker exec` *has* to happen. And this means that **exec will never be equivalent of opening bash and typing commands there**. 

The only question is how this conversion happens. `cmd="$@"` is a most obvious direct array to string conversion. But it will not properly maintain parameters with spaces.   
This conversion will make `exec` think that all spaces are parameter delimiters, which is not true in case of `drush si --site-name="My Site"`. The direct conversion would fail with this line.

Which is why there is a special escaping in  `exec`, that properly escapes spaces, so that `fin exec nano "/var/www/my file"` would actually edit "my file", not "my", which it would edit in case with direct conversion. 

But in `run-cli` there was no such escaping present, because `run-cli` was initially supposed to be used for simple commands. `run-cli` was actually _inferior_ to `exec`, not superior.  
`fin run-cli nano "/var/www/my file"` in fin 1.60.x erroneously edits "my" instead of "my file" because of that.

And I underline that cases when users need this escaping are much more common, then cases when users don't need it.

**TL;DR of part 1: there is special space escaping happening in `exec`, it is intentional and serves regular user convenience.**

* * *

Part 2.

At the same I wanted to give users an ability to run complex commands like `source ~/.profile && npm something`. Of course in this case you *would* need to encapsulate the whole line into a string and feed it to `exec` (just like you would do with `docker exec`). 

But space escaping that helped us in part 1 will actually impede this use-case. It will make `exec` think that you wanted to execute a binary that is literally named `source ~/.profile && npm something`. 

To mitigate it I made a special if clause, that if `exec` only has one parameter, then a special escaping would not happen, and the inferior simple conversion `cmd="$@"` would be made instead. The thinking was that almost nobody would overcomplicate a complex quoted string that is fed to exec and simple conversion would not fail. 

In your example you are trying to use this special clause, but it is failing, because you used `$@` in the string. Bash invisibly converts `"$@"` into space-and-quotes-delimited string, so `exec` thinks that your command is not a single-string command but an array, an if-clause is not applied. 

A workaround that I advised above will help `exec` to see the whole line as a single string.

**TL;DR of part 1 and 2: since there is special escaping happening in `exec`, some rare use-cases that work in `docker exec` will not work in `fin exec`.**

---

Conclusion.

While I could add an option to forcibly prevent space escaping or think about other clever ways to guess whether to apply, I guess it is not really worth it, since one can easily force it by building command string first and then feeding it into `exec`.

The only thing I can/should do is to actually include special escaping into `run-cli`, because currently it does not use it, while it should. And that is what I did in https://github.com/docksal/docksal/commit/044a9912f4401f76fe1e7661286e0090cd2b1163. 

Please use the workaround I described above to achieve the results you need.

**Comment by GMTA** [comment]
Hi @achekulaev,

Thank you for this extremely detailed explanation! I'm going to think about alternative solutions.

> I also need to add cd to current dir at the beginning of that string

Is there a reason why you can't use `docker exec --workdir "${dir}" [..]` for this?

**Comment by achekulaev** [comment]
@GMTA yes there is workdir now. I can use that and I will refactor `exec` to use it, but this does not change everything else.

**Comment by GMTA** [comment]
@achekulaev That's great, and I know that's not going to help you with the other stuff.

I did some testing and found something that might be of interest. `bash -c` is documented as "If the -c option is present, then commands are read from string. If there are arguments after the string, they are assigned to the positional parameters, starting with $0."

This means that if you do this:

`bash -c 'command one two three' command one two three`

The values of "one", "two" and "three" are actually assigned to $1, $2 and $3. I wrote a small test:

`_test() { docker exec -ti -w / docksal-ssh-agent bash -ilc "${*}" "${@}" }`

Now all the following statements work as expected:

```
_test ls
_test ls -al
_test "ls -al"
```

The only dubious thing I found until now was that variable replacement doesn't quite work as you'd expect:
```
_test echo '$SHELL'
# This outputs my shell whereas just running "echo '$SHELL'" gives me "$SHELL"
```

This is because the string `$SHELL` is passed as an argument to `echo` which then interprets it again, unaware of the original quotes being used. Currently `fin exec` handles this case correctly so we might need to improve the snippet a bit.

**Comment by GMTA** [comment]
I fixed the positional arguments provided to `bash` which fixes all the cases I can think of:

```
_test () {
    cmd="${1}"
    args=`printf "%q " "${@:2}"`
    docker exec -ti -w / docksal-ssh-agent bash -ilc "${cmd} ${args}" ${@}
}
```

Test cases I ran:
```
_test echo Testing 1, 2, 3
# Outputs: Testing 1, 2, 3
_test echo "Testing 1, 2, 3"
# Outputs: Testing 1, 2, 3

_test ls
# Gives me a simple listing

_test ls -l
_test "ls -l"
# Both give me a detailed listing

_test echo $SHELL
# Outputs
```

**Comment by achekulaev** [comment]
@GMTA 
Pretty close but not there yet

```bash
#!/bin/bash

_test () {
    cmd="${1}"
    args=`printf "%q " "${@:2}"`
    docker exec -ti -w / docksal-ssh-agent bash -ilc "${cmd} ${args}" ${@}
}

_test 'echo "hello, world" > "/tmp/my file"; cat "/tmp/my file"'
```

Output:

```
§ ./test.sh 
hello, world
cat: can't open '': No such file or directory
```

With `fin exec`:

```bash
§ fin exec 'echo "hello, world" > "/tmp/my file"; cat "/tmp/my file"'
hello, world
```

**Comment by GMTA** [comment]
@achekulaev That is strange. I copied your "_test" literally and it just works for me:

```
$ _test 'echo "hello, world" > "/tmp/my file"; cat "/tmp/my file"'
hello, world
```

**Comment by achekulaev** [comment]
@GMTA I think OS differences. Another problem when doing bash. But I think I got it

This works fine:

```bash
_test () {
    cmd="${1}"
    if [[ -n $2 ]]; then
        args=`printf "%q " "${@:2}"`
    fi
    docker exec -ti -w / docksal-ssh-agent bash -ilc "${cmd} ${args}" ${@}
}
```

I will need to test it with more sophisticated cases which we have, but it looks very promising, thanks!

**Comment by GMTA** [comment]
Just a remark: `[[ -n $2 ]]` checks if the second argument is not an empty string. You might want to change that to something like `[ $# -gt 1 ]` or similar.

**Comment by achekulaev** [comment]
@GMTA isn't it effectively the same?

**Comment by GMTA** [comment]
@achekulaev 

```
#!/bin/bash
set -eo pipefail

_countArgs () {
    echo "Number of arguments: $#"
    [[ -n ${2} ]] && echo "Also, second argument is empty."
}

_countArgs 1 2 3
_countArgs 1 '' 3
```

Output:

```
Number of arguments: 3
Also, second argument is empty.
Number of arguments: 3
```

**Comment by GMTA** [comment]
@achekulaev Not completely, empty arguments are possible:

```
#!/bin/bash
set -eo pipefail

_countArgs () {
    echo "Number of arguments: $#"
    [[ -n ${2} ]] && echo "Also, second argument is not empty."
}

_countArgs 1 2 3
_countArgs 1 '' 3
```

Output:

```
Number of arguments: 3
Also, second argument is not empty.
Number of arguments: 3
```

**Comment by achekulaev** [comment]
@GMTA Your point is correct, however I have literally never seen a command line utility that would accept empty arguments like this.
